plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c

```diff
@@ -0,0 +1,272 @@
+ * @self: a #FuUefiDbxSnapdNotifier
+ *
+ * Notify snapd of that the DBX manager has started.
+ *
+ * Returns: TRUE if the notification was successful.
+ *
+ **/
+gboolean
+fu_uefi_dbx_snapd_notifier_dbx_manager_startup(FuUefiDbxSnapdNotifier *self, GError **error)
+{
+	const char *startup_msg = "{"
+				  "\"action\":\"efi-secureboot-update-startup\""
+				  "}";
+
+	g_debug("snapd notify secureboot manager startup");
+
+	if (!fu_uefi_dbx_snapd_notifier_simple_req(self,
+						   "/v2/system-secureboot",
+						   startup_msg,
+						   strlen(startup_msg),
+						   error)) {
+		g_prefix_error(error, "failed to notify snapd of startup: ");
+		return FALSE;
+	}
+
+	g_debug("snapd notified of secureboot manager startup");
+
+	return TRUE;
+}
+
+/**
+ * fu_uefi_dbx_snapd_notifier_dbx_update_prepare:
+ * @self: a #FuSnapdNotifier
+ * @fw_payload: payload used for the update
+ *
+ * Notify of an upcoming update to the DBX. A successful call shall initiate a
+ * change tracking an update to the DBX on the snapd side.
+ *
+ * Returns: TRUE if the notification was successful.
+ *
+ **/
+gboolean
+fu_uefi_dbx_snapd_notifier_dbx_update_prepare(FuUefiDbxSnapdNotifier *self,
+					      GBytes *fw_payload,
+					      GError **error)
+{
+	gsize bufsz = 0;
+	const guint8 *buf = g_bytes_get_data(fw_payload, &bufsz);
+	g_autofree gchar *b64data = g_base64_encode(buf, bufsz);
+	g_autofree gchar *msg = g_strdup_printf("{"
+						"\"action\":\"efi-secureboot-update-db-prepare\","
+						"\"key-database\":\"DBX\","
+						"\"payload\":\"%s\""
+						"}",
+						b64data);
+
+	g_debug("snapd notify prepare, with %" G_GSIZE_FORMAT " bytes of data", bufsz);
+
+	if (!fu_uefi_dbx_snapd_notifier_simple_req(self,
+						   "/v2/system-secureboot",
+						   msg,
+						   strlen(msg),
+						   error)) {
+		g_prefix_error(error, "failed to notify snapd of prepare: ");
+		return FALSE;
+	}
+
+	g_debug("snapd notified of prepare");
+
+	return TRUE;
+}
+
+/**
+ * fu_uefi_dbx_snapd_notifier_dbx_update_cleanup:
+ * @self: a #FuSnapdNotifier
+ *
+ * Notify of an completed update to one of secureboot key databases. A
+ * successful call shall result in completion of a corresponding change on the
+ * snapd side.
+ *
+ * Returns: TRUE if the notification was successful.
+ *
+ **/
+gboolean
+fu_uefi_dbx_snapd_notifier_dbx_update_cleanup(FuUefiDbxSnapdNotifier *self, GError **error)
+{
+	const char *msg = "{"
+			  "\"action\":\"efi-secureboot-update-db-cleanup\""
+			  "}";
+
+	g_debug("snapd notify cleanup");
+
+	if (!fu_uefi_dbx_snapd_notifier_simple_req(self,
+						   "/v2/system-secureboot",
+						   msg,
+						   strlen(msg),
+						   error)) {
+		g_prefix_error(error, "failed to notify snapd of cleanup: ");
+		return FALSE;
+	}
+
+	g_debug("snapd notified of cleanup");
+
+	return TRUE;
+}
+
+#ifdef FU_UEFI_DBX_SELF_TEST
+const char *
+fu_uefi_dbx_self_test_snapd_notifier_replace_socket_path(const char *path) /* nocheck:name */
```

Comment:
rather than this function, how about sending it from an env variable?

Reply:
changed the tests to use FWUPD_SNAPD_SNAP_SOCKET instead